src/main/kotlin/io/sc3/peripherals/posters/PosterItem.kt

```diff
@@ -0,0 +1,217 @@
+      matrices.translate(tooltipX + 3.0, tooltipY + 3.0, 500.0)
+      matrices.scale(scale, scale, 1f)
+      RenderSystem.enableBlend()
+      DrawableHelper.drawTexture(matrices, -pad, -pad, 0f, 0f, size, size, size, size)
+      val buffer: BufferBuilder = Tessellator.getInstance().buffer
+      val immediateBuffer = VertexConsumerProvider.immediate(buffer)
+      PosterRenderer.draw(matrices, immediateBuffer, posterId, posterState, 240)
+      immediateBuffer.draw()
+      matrices.pop()
+    }
+
+    override fun getHeight(): Int {
+      val posterId = getPosterId(stack)
+      val posterState = getPosterState(posterId, world)
+      return if (posterState != null) 75 else 0
+    }
+
+    override fun getWidth(textRenderer: TextRenderer) = 72
+  }
+
+  companion object {
+    const val POSTER_KEY = "poster"
+
+    fun getPosterName(posterId: String) = "posters/${posterId.take(2)}/$posterId"
+
+    fun create(world: World, data: PosterPrintData): ItemStack {
+      val itemStack = ItemStack(ModItems.poster)
+      if (data.posterId != null) {
+        setPosterId(itemStack, data.posterId!!, data)
+      } else {
+        createPosterState(itemStack, world, data)
+      }
+
+      return itemStack
+    }
+
+    private fun setPosterId(stack: ItemStack, id: String, data: PosterPrintData) {
+      stack.orCreateNbt.putString(POSTER_KEY, id)
+      stack.nbt!!.copyFrom(data.toItemNbt())
+    }
+
+    private fun createPosterState(
+      stack: ItemStack,
+      world: World,
+      data: PosterPrintData
+    ): PosterState {
+      val (i, state) = allocatePosterId(world, data)
+      setPosterId(stack, i, data)
+      return state
+    }
+
+    fun allocatePosterId(
+      world: World, data: PosterPrintData
+    ): Pair<String, PosterState> {
+      val posterState = PosterState().also {
+        for (x in 0 until 128) {
+          for (y in 0 until 128) {
+            it.setColor(x, y, data.colors[x + y * 128].toByte())
+          }
+        }
+
+        for (i in 0 until data.palette.size) {
+          it.palette[i] = data.palette[i]
+        }
+      }
+
+      val i = allocateHash(world, posterState)
+
+      world.putPosterState(getPosterName(i), posterState)
+      return i to posterState
+    }
+
+    private fun allocateHash(world: World, data: PosterState): String {
+      val contents = ByteArrayOutputStream().use { baos ->
+        DataOutputStream(baos).use { dos ->
+          NbtIo.write(data.writeNbt(NbtCompound()), dos)
+          baos.toByteArray()
+        }
+      }
+
+      var disambiguation = 0
+      while (true) {
+        val digest = MessageDigest.getInstance("SHA-256")
+        val hash = digest.digest(contents + disambiguation.toString().toByteArray())
+        val hex = hash.joinToString("") { "%02x".format(it) }
+
+        val existingData = world.getPosterState(getPosterName(hex))
+        if (existingData == null || existingData == data) {
+          return hex
+        }
+
+        disambiguation += 1
```

Comment:
This should never happen, but can you put an upper limit here (of something reasonable like a po2 >=128) and make it throw an exception if it's still going?